keyboard/_darwinmouse.py

```diff
@@ -0,0 +1,173 @@
+        self.blocking = blocking
+        self.callback = callback
+        self.listening = True
+
+    def run(self):
+        """ Creates a listener and loops while waiting for an event. Intended to run as
+        a background thread. """
+        self.tap = Quartz.CGEventTapCreate(
+            Quartz.kCGSessionEventTap,
+            Quartz.kCGHeadInsertEventTap,
+            Quartz.kCGEventTapOptionDefault,
+            Quartz.CGEventMaskBit(Quartz.kCGEventLeftMouseDown) |
+            Quartz.CGEventMaskBit(Quartz.kCGEventLeftMouseUp) |
+            Quartz.CGEventMaskBit(Quartz.kCGEventRightMouseDown) |
+            Quartz.CGEventMaskBit(Quartz.kCGEventRightMouseUp) |
+            Quartz.CGEventMaskBit(Quartz.kCGEventOtherMouseDown) |
+            Quartz.CGEventMaskBit(Quartz.kCGEventOtherMouseUp) |
+            Quartz.CGEventMaskBit(Quartz.kCGEventMouseMoved) |
+            Quartz.CGEventMaskBit(Quartz.kCGEventScrollWheel),
+            self.handler,
+            None)
+        loopsource = Quartz.CFMachPortCreateRunLoopSource(None, self.tap, 0)
+        loop = Quartz.CFRunLoopGetCurrent()
+        Quartz.CFRunLoopAddSource(loop, loopsource, Quartz.kCFRunLoopDefaultMode)
+        Quartz.CGEventTapEnable(self.tap, True)
+
+        while self.listening:
+            Quartz.CFRunLoopRunInMode(Quartz.kCFRunLoopDefaultMode, 5, False)
+
+    def handler(self, proxy, e_type, event, refcon):
+        # TODO Separate event types by button/wheel/move
+        scan_code = Quartz.CGEventGetIntegerValueField(event, Quartz.kCGKeyboardEventKeycode)
+        key_name = name_from_scancode(scan_code)
+        flags = Quartz.CGEventGetFlags(event)
+        event_type = ""
+        is_keypad = (flags & Quartz.kCGEventFlagMaskNumericPad)
+        if e_type == Quartz.kCGEventKeyDown:
+            event_type = "down"
+        elif e_type == Quartz.kCGEventKeyUp:
+            event_type = "up"
+
+        if self.blocking:
+            return None
+
+        self.callback(KeyboardEvent(event_type, scan_code, name=key_name, is_keypad=is_keypad))
+        return event
+
+# Exports
+
+def init():
+    """ Initializes mouse state """
+    pass
+
+def listen(queue):
+    """ Appends events to the queue (ButtonEvent, WheelEvent, and MoveEvent). """
+    if not os.geteuid() == 0:
+        raise OSError("Error 13 - Must be run as administrator")
+    listener = MouseEventListener(lambda e: queue.put(e) or is_allowed(e.name, e.event_type == KEY_UP))
+    t = threading.Thread(target=listener.run, args=())
+    t.daemon = True
+    t.start()
+
+def press(button=LEFT):
+    """ Sends a down event for the specified button, using the provided constants """
+    location = get_position()
+    button_code, button_down, _, _ = _button_mapping[button]
+    e = Quartz.CGEventCreateMouseEvent(
+        None,
+        button_down,
+        location,
+        button_code)
+
+    # Check if this is a double-click (same location within the last 300ms)
```

Comment:
I'm not a big fan of automatically detecting double clicks, specially like this. The OS likely has internal settings to adjust sensitivity and timing. Also, the logic is just complex enough to potentially hide bugs.

But since the abstract code is at fault here ( https://github.com/boppreh/keyboard/blob/master/keyboard/mouse.py#L51 ), I realize the fix is a bit more involved. The ideal solution would be to always specify the type of event, both when listening and when sending, so there's never any ambiguity. I'll that to my TODO list.